compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -75,132 +75,280 @@ std::optional<uint32_t> getNumberIterations(scf::ForOp loop) {
   }
 }
 
-/// Computes the number of BD IDs required between the current
-/// DMA copy operation and its corresponding DMA wait operation.
-/// If a sub-loop is encountered, it assumes that a BD ID expression will be
-/// used within the sub-loop. This approach ensures that the inner loop has
-/// access to a greater number of BD IDs, which is favorable for enabling
-/// efficient BD chaining in subsequent passes.
-///
-/// Example:
-/// %0 = dma_copy {bd_id = 0}   // Current DMA copy
-/// scf.for %arg0 = %c0 to %c1 step %c2 {
-///   %1 = dma_copy {bd_id = %arg0 + 1}  // DMA copy inside a sub-loop
-///   dma_wait(%1)                       // Wait for the sub-loop DMA copy
-/// }
-/// dma_wait(%0)   // Current DMA wait
-///
-/// In this example:
-/// - The current DMA copy (%0) requires 1 BD ID.
-/// - The sub-loop executes 2 iterations, each requiring 1 BD ID.
-/// - Therefore, the required number of BD IDs is:
-///  3 = 1 (current) + 2 * 1(sub-loop).
-void getNumRequiredBdIds(
-    scf::ForOp loop, AMDAIE::NpuDmaCpyNdOp currDmaOp, AMDAIE::TileOp tileOp,
-    DenseMap<Value, ChannelBdIdGenerator> &shimTileToGeneratorMap,
-    uint32_t &bdIdCount) {
-  bool startCounting =
-      (currDmaOp == nullptr);  // Start immediately if no currDmaOp
-
-  for (auto &op : loop.getOps()) {
-    if (auto npuDmaOp = dyn_cast<AMDAIE::NpuDmaCpyNdOp>(op)) {
-      // Skip until currDmaOp is found
-      if (npuDmaOp == currDmaOp) startCounting = true;
-      if (!startCounting) continue;
-      if (npuDmaOp.getSource()) {
+/// A DmaBatch contains the following :-
+///   1. DmaOps within the current batch.
+///   2. Required Bd Ids by the current batch.
+///   3. Nested DmaBatch chain.
+///   4. Pointer to the sibling DmaBatch.
+///   4. A pointer to the parent op if the parent of is a scf.for.
+typedef struct DmaBatchStruct {
+  SmallVector<AMDAIE::NpuDmaCpyNdOp> currentDmaOps = {};
+  int32_t requiredBdIds = 0;
+  SmallVector<DmaBatchStruct *> immediateInnerBatches = {};
+  DmaBatchStruct *nextDmaBatch = nullptr;
+  scf::ForOp forOpParent = nullptr;
+} DmaBatch;
+
+using TileDmaBatchGraph = llvm::MapVector<AMDAIE::TileOp, DmaBatch *>;
```

Comment:
You might want to encapsulate this as well with a class/struct to add appropriate utility methods to access and update data, like `inferBdIdsRequiredInBatches` below.